latency-breakdown.md

```diff
@@ -0,0 +1,799 @@
+    tikv_scheduler_latch_wait_duration_seconds{type="acquire_pessimistic_lock"} +
+    tikv_storage_engine_async_request_duration_seconds{type="snapshot"} +
+    (lock in-mem key count + lock on-disk key count) * lock read duration +
+    lock on-disk key count / (lock in-mem key count + lock on-disk key count) *
+    lock write duration
+```
+
+From 6.0, TiKV uses [in-mem pessimisitc lock](/pessimistic-transaction.md#in-memory-pessimistic-lock) by default. In-mem pessimistic lock bypass the async write process. The snapshot duration is explained in [TiKV Snapshot section](#tikv-snapshot). `lock write duration` is the duration of writing on-disk lock, which is explained in the [Async Write section](#async-write).
+
+```text
+lock in-mem key count =
+    sum(rate(tikv_in_memory_pessimistic_locking{result="success"})) /
+    sum(rate(tikv_grpc_msg_duration_seconds_count{type="kv_pessimistic_lock"}}))
+
+lock on-disk key count =
+    sum(rate(tikv_in_memory_pessimistic_locking{result="full"})) /
+    sum(rate(tikv_grpc_msg_duration_seconds_count{type="kv_pessimistic_lock"}}))
+```
+
+The count of in-mem and on-disk locked keys can be calculated by in-mem lock counter.
+
+```text
+lock read duration(from disk) =
+    sum(rate(tikv_storage_rocksdb_perf{metric="block_read_time",req="acquire_pessimistic_lock"})) / sum(rate(tikv_storage_rocksdb_perf{metric="block_read_count",req="acquire_pessimistic_lock"}))
+```
+
+TiKV reads the value of the keys before acquiring the locks, the read duration can be calculated by RocksDB perf context.
+
+### Commit
+
+```railroad+diagram
+Diagram(
+    Stack(
+        Sequence(
+            Choice(
+                0,
+                Comment("use 2pc or causal consistency"),
+                NonTerminal("Get min-commit-ts"),
+            ),
+            Optional("Async prewrite binlog"),
+            NonTerminal("Prewrite mutations"),
+            Optional("Wait prewrite binlog result"),
+        ),
+        Sequence(
+            Choice(
+                1,
+                Comment("1pc"),
+                Sequence(
+                    Comment("2pc"),
+                    NonTerminal("Get commit-ts"),
+                    NonTerminal("Check schema, try to amend if needed"),
+                    NonTerminal("Commit PK mutation"),
+                ),
+                Sequence(
+                    Comment("async-commit"),
+                    NonTerminal("Commit mutations asynchronously"),
+                ),
+            ),
+            Choice(
+                0,
+                Comment("committed"),
+                NonTerminal("Async cleanup"),
+            ),
+            Optional("Commit binlog"),
+        ),
+    ),
+)
+```
+
+```text
+commit =
+    Get_latest_ts_time +
+    Prewrite_time +
+    Get_commit_ts_time +
+    Commit_time
+
+Get_latest_ts_time = Get_commit_ts_time =
+    pd_client_cmd_handle_cmds_duration_seconds{type="wait"}
+
+prewrite_ratio = ceil(
```

Comment:
What does the "ratio" mean here..? I guess you mean the amplification _factor_?

Personally, I would call it "round"?